### PR TITLE
Support data streams in enrich policy indices

### LINKED ISSUE #98836: [Enrich] Improve error returned when data stream targeted as index
### Description

If a datastream is targeted by a enrich policy like the following:

```
PUT /_enrich/policy/my-policy
{
  "match": {
    "indices": ["data_stream"],
    "match_field": "fieldA",
    "enrich_fields": ["fieldB", "fieldC"]
  }
}
```

then an error like the following is returned:

```json
{
  "error": {
    "root_cause": [
      {
        "type": "index_not_found_exception",
        "reason": "no such index [data_stream]",
        "resource.type": "index_or_alias",
        "excluded_ds": "true",
        "resource.id": "data_stream",
        "index_uuid": "_na_",
        "index": "data_stream"
      }
    ],
    "type": "index_not_found_exception",
    "reason": "no such index [data_stream]",
    "resource.type": "index_or_alias",
    "excluded_ds": "true",
    "resource.id": "data_stream",
    "index_uuid": "_na_",
    "index": "data_stream"
  },
  "status": 404
}
```

Other that decypher the combination of `"reason": "no such index [data_stream]"` and `"excluded_ds": "true"` it's not easy to understand that data streams are not supported. Also the documentation does not mention anything about data streams as well.

It would be nice to have a better error message in this case that explains that the provided `index` is a data stream and is not supported.

### PR DESCRIPTION
**Pull request by dakrone**
Previously when someone tried to use a data stream in the `indices` keyword for an enrich policy, an exception would be thrown. This commit changes the resolution so that data streams can be used.

Resolves #98836

### CODE REVIEW

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/enrich/50_data_stream.yml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  - do:
+      ingest.delete_pipeline:
+        id: test_pipeline
+
+  - do:
+      enrich.delete_policy:
+        name: test_policy
+
+---
+"enrich documents over _bulk via a data stream":
```
Comment (nielsbauman) [inline x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/enrich/50_data_stream.yml:61]: Would we also want a test that ensures that the enrich policy is able to fetch data from non-write index backing indices? Or is that overkill?

I also don't see any YAML REST tests (or integration tests in general for that matter) that cover using aliases in enrich. Is that something worth adding if we have a specific test for data streams as well?

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichStore.java [hunk 2]
```diff
@@ -86,6 +86,7 @@ public static void putPolicy(
                 String[] concreteIndices = indexNameExpressionResolver.concreteIndexNames(
                     current,
                     IndicesOptions.strictExpandOpen(),
+                    true,
```
Comment (nielsbauman) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichStore.java:89]: This is not a blocker for this PR, but I'm curious to hear what your take is on Martijn's [comment](https://github.com/elastic/elasticsearch/issues/98836#issuecomment-1692650512) about possible performance issues when it comes to using aliases and data streams in enrich. I would think this is a pretty realistic issue, but maybe I'm missing something. I don't see any tip or warning [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/enrich-setup.html) or [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/ingest-enriching-data.html), maybe it's worth adding one over there?

**Review by nielsbauman — COMMENTED** [review]
This implementation was more straightforward than I expected 😄. I do have 2 semi-discussion comments, one of which is definitely not a blocker.

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichStore.java (hunk 2 above)
Comment (dakrone) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichStore.java:89]: I do think it's worth documenting that enrich indices are not expected to be large. I'll do that in a follow-up PR. As for Martijn's comment, yes, unfortunately large amounts of data probably isn't the best idea for enrich indices, but we already have that issue with allowing wildcard patterns (someone could use `logs-*` and grab hundreds of indices already), so I don't think we're making anything worse here.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/enrich/50_data_stream.yml (hunk 1 above)
Comment (dakrone) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/enrich/50_data_stream.yml:61]: That's a good idea, I've added a yaml test for an enrich policy that uses an alias.

**Comment by dakrone** [comment]
@elasticmachine update branch

**Review by nielsbauman — APPROVED** [review]
LGTM! Thanks for adding the test and thanks in advance for the docs follow-up :)

**Comment by nicpenning** [comment]
Will that also resolve this: https://github.com/elastic/kibana/issues/178829 ?

**Pull request merged**